### PR TITLE
GROUP BY Continuation Token

### PR DESCRIPTION
**Pull request by bchong95**
# GROUP BY Continuation Token

## Description

This PR adds continuation token support for GROUP BY. The continuation token is similar to AGGREGATES in the sense that aggregates can resume a single grouping. This PR adds support for multiple groupings. 

The continuation token will look like:

```c#
private struct GroupByContinuationToken
        {
            public GroupByContinuationToken(
                string groupingTableContinuationToken,
                string sourceContinuationToken,
                int numPagesDrainedFromGroupingTable)
            {
                this.GroupingTableContinuationToken = groupingTableContinuationToken;
                this.SourceContinuationToken = sourceContinuationToken;
                this.NumPagesDrainedFromGroupingTable = numPagesDrainedFromGroupingTable;
            }
      }
```

And the grouping table continuation token is just a table of the hash of the grouping keys followed by it's corresponding single grouping continuation token.

This enables running group by queries with continuation tokens. The feature is needed for query compute gateway. 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Json/JsonWriter.JsonTextWriter.cs [hunk 1]
```diff
@@ -398,20 +398,11 @@ private void PrefixMemberSeparator()
 
             private bool RequiresEscapeSequence(char value)
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Json/JsonWriter.JsonTextWriter.cs]: >private [](start = 12, length = 7)

please revert

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.Client.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+                    this.AggregateGroupings(sourceResponse.CosmosElements);
+
+                    this.isDone = false;
+                }
+
+                // Stage 2:
+                // Emit the results from the grouping table page by page
+                IEnumerable<SingleGroupAggregator> groupByValuesList = this.groupingTable
+                    .OrderBy(kvp => kvp.Key)
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.Client.cs]: >.OrderBy(kvp => kvp.Key) [](start = 20, length = 24)

consider not doing this

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.Compute.cs [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+                    jsonWriter.WriteFieldName(kvp.Key.ToString());
+                    jsonWriter.WriteStringValue(kvp.Value.GetContinuationToken());
+                }
+                jsonWriter.WriteObjectEnd();
+
+                string result = Utf8StringHelpers.ToString(jsonWriter.GetResult());
+                return result;
+            }
+
+            private readonly struct GroupByContinuationToken
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.Compute.cs]: >struct [](start = 29, length = 6)

class

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.Compute.cs [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+                {
+                    continuationToken = default;
+                    return false;
+                }
+                if (this.Source.IsDone)
+                {
+                    continuationToken = new GroupByContinuationToken(
+                        this.GetGroupingTableContinuationToken(),
+                        GroupByDocumentQueryExecutionComponent.DoneReadingGroupingsContinuationToken,
+                        this.numPagesDrainedFromGroupingTable).ToString();
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.Compute.cs]: skip the groups that have already been returned

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.Compute.cs [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+                        responseLengthBytes: sourceResponse.ResponseLengthBytes);
+
+                    this.isDone = false;
+                }
+                else
+                {
+                    // Stage 2:
+                    // Emit the results from the grouping table page by page
+                    IEnumerable<SingleGroupAggregator> groupByValuesList = this.groupingTable
+                        .OrderBy(kvp => kvp.Key)
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.Compute.cs]: >.OrderBy(kvp => kvp.Key) [](start = 24, length = 24)

no need for this

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.Compute.cs [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+
+                private DoneDocumentQueryExecutionComponent()
+                {
+                }
+
+                public bool IsDone => true;
+
+                public void Dispose()
+                {
+                    throw new NotImplementedException();
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.Compute.cs]: >throw new NotImplementedException(); [](start = 20, length = 36)

think of better default implementation

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.cs [hunk 7]
```diff
@@ -71,128 +73,111 @@ internal sealed class GroupByDocumentQueryExecutionComponent : DocumentQueryExec
...
+                {
+                    throw this.cosmosQueryClient.CreateBadRequestException($"Invalid GroupingTableContinuationToken");
+                }
+
+                foreach (KeyValuePair<string, CosmosElement> kvp in parsedGroupingTableContinuations)
+                {
+                    string key = kvp.Key;
+                    CosmosElement value = kvp.Value;
+
+                    UInt192 groupByKey = UInt192.Parse(key);
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.cs]: >UInt192 [](start = 20, length = 7)

128

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.cs [hunk 8]
```diff
@@ -71,128 +73,111 @@ internal sealed class GroupByDocumentQueryExecutionComponent : DocumentQueryExec
...
+            this.groupingTable = new Dictionary<UInt192, SingleGroupAggregator>();
 
             this.cosmosQueryClient = cosmosQueryClient;
-            this.groupingTable = new Dictionary<UInt192, SingleGroupAggregator>();
             this.groupByAliasToAggregateType = groupByAliasToAggregateType;
             this.orderedAliases = orderedAliases;
             this.hasSelectValue = hasSelectValue;
+            this.numPagesDrainedFromGroupingTable = numPagesDrainedFromGroupingTable;
+
+            if (groupingTableContinuationToken != null)
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.cs]: >f (groupingTableContinuationToken != null) [](start = 13, length = 42)

move this code to compute

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.cs [hunk 9]
```diff
@@ -71,128 +73,111 @@ internal sealed class GroupByDocumentQueryExecutionComponent : DocumentQueryExec
...
+                    string key = kvp.Key;
+                    CosmosElement value = kvp.Value;
+
+                    UInt192 groupByKey = UInt192.Parse(key);
+
+                    if (!(value is CosmosString singleGroupAggregatorContinuationToken))
+                    {
+                        throw this.cosmosQueryClient.CreateBadRequestException($"Invalid GroupingTableContinuationToken");
+                    }
+                    SingleGroupAggregator singleGroupAggregator = SingleGroupAggregator.Create(
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.cs]: >SingleGroupAggregator singleGroupAggregator = SingleGroupAggregator.Create( [](start = 20, length = 75)

new line

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.Compute.cs [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+            private string GetGroupingTableContinuationToken()
+            {
+                IJsonWriter jsonWriter = JsonWriter.Create(JsonSerializationFormat.Text);
+                jsonWriter.WriteObjectStart();
+                foreach (KeyValuePair<UInt192, SingleGroupAggregator> kvp in this.groupingTable)
+                {
+                    jsonWriter.WriteFieldName(kvp.Key.ToString());
+                    jsonWriter.WriteStringValue(kvp.Value.GetContinuationToken());
+                }
+                jsonWriter.WriteObjectEnd();
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.Compute.cs]: array of tuples and avoid redundant property names
use binary encoding

Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/ExecutionComponent/GroupByDocumentQueryExecutionComponent.Compute.cs]: Group BY continuation token is a composite of the aggregate continuation token, so the property names will have to be repeated unless we design a new special continuation token just for GROUP BY + aggregates. 

---
In reply to: [341783873](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/866#discussion_r341783873) [](ancestors = 341783873)

File: Microsoft.Azure.Cosmos/src/Query/Core/UInt128.cs [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+            }
+
+            byte[] bytes = new byte[UInt128.Length];
+            for (int index = 0; index < UInt128.Length; index++)
+            {
+                bytes[index] = byte.Parse(hexPairs[index], System.Globalization.NumberStyles.HexNumber, null);
+            }
+
+            return UInt128.FromByteArray(bytes);
+        }
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/UInt128.cs]: extension method?

**Pull request merged**